성적관리 [STEP 2] wily.kim

wilybear opened the pull request:
@wonhee009 

안녕하세요. 원희님 STEP2 PR을 보냅니다!
우선 코드 설명에 앞서서 시간이 부족하여 생각만 하고 수정을 못하거나 고민을 많이 하지 못한 부분들이 있습니다. 열심히 리뷰해주시는데 죄송하네요😭. 기능 구현은 모두 완료 되었으나 구조적으로 미숙합니다.

### GradeManger.swift
학생추가 삭제 기능 외에 성적 추가(변경), 삭제, 평균 보기 기능을 추가하였습니다. 이 과정에서 각 프로세스에서 발생할 수 있는 결과를 프린트 하기 위해서 단순히 success 성공여부에 따라 결과값을 콘솔에서 출력하였습니다. 
추후에 다양한 상황이 발생할 수 있음을 깨달았습니다. 예를 들어 성적 삭제 시에 존재하지 않은 학생의 성적을 지우려거나 학생이 가지고 있지 않은 과목의 성적을 지우려거나 하는 상황이 발생할 수 있었습니다. 
따라서 다양한 결과값을 출력하기 위해서 `deleteGradeProcess()` 부터는 다양한 error를 만들어 상황에 맞게 에러를 던지고 받아서 출력하는 형식으로 수정하였습니다. 그 이전에 부분들을 시간이 부족하여 수정하지 못하였고 에러 또한 `ResultError`, `InputError`, etc 로 나누고 싶었지만.. 나누지 못했습니다.

### GradeConsoleMessage.swift
이 파일 안에는 String으로 `Console`에서 존재하는 메시지들을 enum으로 만들려고 했습니다. String으로 `Console`파일에 남아있는 것은 좋지 않은 것이라고 생각하여 분리하였지만 코드가 조금더 더러워진거 같아서 고민입니다.

### Menu.swift
기존에 `Menu` enum이 존재했던 파일입니다. `Menu` enum에 `menuTest`라는 String을 추가하여 각 메뉴에 알맞는 String들을 넣어서 `GradeConsole`의 `showMenu` 함수를 간단하게 만들어보았습니다.

### Student.swift
`Grade` Enum을 추가하였고 성적의 경우 A+ ~ F까지 정해져있기 때문에 하나씩 입력했습니다.

### GradeConsole.swift
각 기능 마다 `Input`함수와 `validate` 함수를 제작하여 넣었습니다. 코드의 가독성과 유지보수를 쉽게 하기 위해서 String을 없애보려고 했지만 시간이 없어서 전부 없애지 못하였고 Result 부분 (ex `showAddStudentResult`) 같은 경우는 더 보기 안좋아진거 같습니다..

### 고민되는 부분
위에서 애기한 부분외에는:
1. STEP1에서 STEP2로 진행하면서 기능들이 추가되고 `GradeConsole`과 `GradeManager`들의 파일크기가 점점 커져갔습니다. 파일을 열때마다 찾는게 점점 힘들어지는데 이 클래스와 구조체를 쪼갤 수 있는 방법에는 뭐가 있을까요??

wilybear committed:
feat: STEP2 성적 추가(변경) 기능 추가, Grade Enum추가

Menu 관련 스트링 MenuConsole에서 -> Menu Enum으로
성적 추가(변경) 기능 추가
 	Grade Enum 추가
	성적 추가 입력에 대한 검증 추가
	성적관련 검증 함수 테스트 추가

wilybear committed:
feat: STEP2 학생 성적 삭제 기능 추가

학생 성적 삭제 기능 추가
시간이 된다면 Failed한 Result들은 -> Error처리로 받는게 좋을듯

wilybear committed:
feat: STEP2 평점보기 기능 추가

평점 보기 기능 추기

wonhee009 commented on GradeManager/GradeConsole.swift:
```diff
@@ -22,6 +22,30 @@ struct GradeConsole {
     }
+    
+    func getAddGradeInput() throws -> (String, String, Grade) {
+        guard let input = readLine(), validateStudentGradeInputForAddition(input) else {
```
```suggestion
        guard let input = readLine(), 
                   validateStudentGradeInputForAddition(input) else {
```
너무 길면 이렇게 줄바꿈을 하는 것도 가독성을 올리는 방법 중 하나입니다☺️

wonhee009 commented on GradeManager/GradeConsole.swift:
```diff
@@ -64,6 +95,28 @@ struct GradeConsole {
         return (isNumberic || isAlphabet) && !input.isEmpty
     }
     
+    func validateStudentGradeInputForAddition(_ input: String) -> Bool{
+        guard validate(input) else { return false }
+        let inputs = input.components(separatedBy: " ")
+        guard inputs.count == 3 else { return false }
+        let grade = inputs[2]
+        guard validateGrade(grade) else {return false}
+        
+        return true
+    }
+    
+    func validateStudentGradeInputForDeletion(_ input: String) -> Bool{
+        guard validate(input) else { return false }
+        let inputs = input.components(separatedBy: " ")
+        guard inputs.count == 2 else { return false }
+        
+        return true
+    }
```
성적 추가를 위한 검증 메서드, 성적 삭제를 위한 검증 메서드로 이해했습니다.
두 메서드의 플로우가 유사하다고 느껴지는데요.
(input 검증 -> 공백을 기준으로 문자열 split -> 필요한 문자열 갯수 확인...)
이런 과정을 공통화 할 수 있지 않을까요?

그리고 공백을 기준으로 자른 문자열 배열에서 학생 이름, 과목명에 대한 검증은 진행하지 않아도 되는걸까요?

wilybear replied:
> GradeConsoleMessage
> 저도 관련 Text를 한 곳에 모아두고 사용하는걸 선호하는 편인데요.
> 이렇게 모아두면 갖는 장점과 단점이 무엇일까요? 같이 얘기해 보는 것도 좋을거 같습니다.

이렇게 Text를 모아두었을 때 제가 생각하는 장점은 같은 Text를 사용하는 경우 수정 시에 GradeConsoleMessage만 수정하면 되므로 실수가 줄어듭니다. 또 Text자체의 의미를 쉽게 파악가능하다고 생각합니다.
단점으로는.. 코드양이 조금 더 많아진다? 정도 있는 것 같습니다.


> (input 검증 -> 공백을 기준으로 문자열 split -> 필요한 문자열 갯수 확인...)
> 이런 과정을 공통화 할 수 있지 않을까요?

메소드를 위에 기능별로 잘게 쪼개서 공통적으로 사용해보았습니다.


> 그리고 공백을 기준으로 자른 문자열 배열에서 학생 이름, 과목명에 대한 검증은 진행하지 않아도 되는걸까요?

주어진 요구사항에서는 이름과 과목명에 대한 입력 제한이 없었고 " " 빈칸으로 split하여 갯수가 일치하게 나오려면 문자가 존재해야 하고 또한 이전에 validate함수에서 알파벳, 숫자로만 이루어져 있는지 체크했기 때문에 따로 검증을 하지 않았습니다.
하지만 제가 생각하기에는 이럴 경우 추후에 검증에 대한 추가사항 요청 시 수정이 어렵고 의도를 보여주기 어렵다고 생각되어 각각 검증함수를 만들고 이름은 알파벳만, 과목이 경우 알파벳과 숫자만으로 검증하였습니다.

wonhee009 replied:
> 이렇게 Text를 모아두었을 때 제가 생각하는 장점은 같은 Text를 사용하는 경우 수정 시에 GradeConsoleMessage만 수정하면 되므로 실수가 줄어듭니다. 또 Text자체의 의미를 쉽게 파악가능하다고 생각합니다.
단점으로는.. 코드양이 조금 더 많아진다? 정도 있는 것 같습니다.

코드의 양이 많아지는 점도 생각해보면 동일하게 반복되는 텍스트를 raw하게 입력하는 것보다는 모아두고 쓰는 편이 더 줄어들 수 있지 않을까요?
저도 말씀해주신 장점들에 대해서는 다 공감합니다!
제가 생각하는 단점은 현재는 동일하게 사용하는 텍스트여도 나중에 스펙이 바뀌면서 일정 부분에서 텍스트가 바뀌게 된다면 다시 분리해줘야한다는 점인거 같습니다:)

> 메소드를 위에 기능별로 잘게 쪼개서 공통적으로 사용해보았습니다.
👍

> 주어진 요구사항에서는 이름과 과목명에 대한 입력 제한이 없었고 " " 빈칸으로 split하여 갯수가 일치하게 나오려면 문자가 존재해야 하고 또한 이전에 validate함수에서 알파벳, 숫자로만 이루어져 있는지 체크했기 때문에 따로 검증을 하지 않았습니다.
하지만 제가 생각하기에는 이럴 경우 추후에 검증에 대한 추가사항 요청 시 수정이 어렵고 의도를 보여주기 어렵다고 생각되어 각각 검증함수를 만들고 이름은 알파벳만, 과목이 경우 알파벳과 숫자만으로 검증하였습니다.

네넵
저는 모든 메뉴에서 학생 이름, 과목명 검증이 반복되는 점에서 해당 검증 함수를 두고, split한 배열의 갯수도 검증하는 함수를 따로 두는 방식을 생각했습니다.
메뉴 실행 함수에서는 필요한 검증 함수들을 조합해서 사용할 수 있는 방향을 생각해보았는데요.
작성해주신 이유들도 타당하다고 생각합니다.

wonhee009 commented on GradeManager/GradeConsole.swift:
```diff
@@ -22,6 +22,30 @@ struct GradeConsole {
         return menu
     }
+    
+    func getAddGradeInput() throws -> (String, String, Grade) {
+        guard let input = readLine(), validateStudentGradeInputForAddition(input) else {
+            throw InputError.invalidInput
+        }
+        let seperatedInput = input.components(separatedBy: " ")
+        let name = seperatedInput[0]
+        let subject = seperatedInput[1]
+        guard let grade = Grade(rawValue: seperatedInput[2]) else {
+            throw InputError.invalidInput
+        }
+        return (name, subject, grade)
+    }
+    
+    func getDeleteGradeInput() throws -> (String, String) {
+        guard let input = readLine(), validateStudentGradeInputForDeletion(input) else {
+            throw InputError.invalidInput
+        }
+        let seperatedInput = input.components(separatedBy: " ")
+        let name = seperatedInput[0]
+        let subject = seperatedInput[1]
+        
+        return (name,subject)
+    }
```
`(String, String, Grade)`, `(String, String)`으로 반환해주고 있네요🤔
이후에 어떤게 name인지 헷갈리지 않을까요?
```swift
func getAddGradeInput() -> (name: String, subject: String, grade: Grade)
```
이렇게 이름을 부여할수도 있습니다☺️

wilybear replied:
이런 방법도 있었군요.. 감사합니다🤩

wonhee009 commented on GradeManager/GradeConsoleMessage.swift:
```diff
@@ -0,0 +1,46 @@
+typealias Name = String
+typealias Subject = String
+
+enum ResultMessage {
```
파일명과 일치하지 않은거 같아요🤔

wonhee009 commented on GradeManager/GradeConsoleMessage.swift:
```diff
@@ -0,0 +1,46 @@
+enum ResultMessage {
+    case addStudentFail(Name)
+    case addStudentSuccess(Name)
+    case noStudentFound(Name)
```
`studentNotFound(name)`은 어떤가요?🤔

wonhee009 commented on GradeManager/GradeConsole.swift:
```diff
@@ -64,6 +95,28 @@ struct GradeConsole {
+        return Grade(rawValue: input) != nil
+    }
+    
     func printInputError(_ error: Error){
         guard let inputError = error as? InputError else { return }
         print(inputError.errorDescription)
```
InputError로 컨버팅해주는 이유가 따로 있을까요?☺️

wonhee009 commented on GradeManager/GradeManager.swift:
```diff
@@ -68,4 +112,29 @@ class GradeManager {
+    }
+    
+    func getStudentElement(by name: String) -> Student? {
+        guard let index = students.firstIndex(where: {$0.name == name}) else { return nil }
```
```swift
students.first(where: { $0.name == name })
```
이 메서드로는 안되나요?🤔

wonhee009 commented on GradeManager/GradeManager.swift:
```diff
@@ -68,4 +112,29 @@ class GradeManager {
+        return true
+    }
+    
+    func getStudentElement(by name: String) -> Student? {
+        guard let index = students.firstIndex(where: {$0.name == name}) else { return nil }
+       return students[index]
+    }
```
`Element`는 없어도 충분히 의미 전달이 가능할거 같습니다!
해당 메서드는 `GradeManager`에서만 사용하고 있는거로 보이는데요.
`private`으로 선언하지 않은 이유가 따로 있을까요?
(혹시 테스트코드 때문에 열어두신거라면 private한 메서드를 테스트 때문에 열어두어야 하나에 대해서 같이 얘기해보면 좋을거 같습니다!👐)

wilybear replied:
이 함수에 대해서는 테스트를 진행하지 않았습니다.
하지만 테스트 코드 작성 관련해서 private를 선언할지 말지 항상 고민하게됩니다.

제가 인터넷에서 찾아본 결과 private method들에 대해서는 테스트를 진행하지 않는 것이 옳다 라고 이해했는데요
하지만 이번 프로젝트에서 제가 작성한 테스트 코드들은 모두 private method들이였고 이를 테스트하기 위해 private를 지웠습니다.
이런 경우에는 어떻게 해야할까요? (ex GradeConsole의 validate 함수들)

wonhee009 replied:
저도 "private method에 대해서는 테스트를 진행하지 않는다"에 대해서 공감합니다.
외부에서 사용하지 않는 부분들은 private으로 해두는게 이후 여러 에러들을 방지할 수 있는 방법이라고 생각하기도 하고 테스트를 위해 이런 부분들을 감수해야하나?라고 생각합니다.
따라서 공통화 할 수 있는 부분들은 공통화하고, 메서드를 가능한한 작게 구성하려고 노력합니다.

wonhee009 commented on GradeManager/GradeManager.swift:
```diff
@@ -58,6 +58,50 @@ class GradeManager {
+        }
+    }
+    
+    private func averageGradeProcess() {
+        do{
+            let name = try gradeConsole.getNameInput()
+            guard let student = getStudentElement(by: name) else {
+                throw InputError.invalidName(name)
+            }
+            let subjects = student.subjects
+            subjects.forEach { subject, grade in
+                self.gradeConsole.showGrade(grade,subject: subject)
+            }
+            let averageGrade = try getAverageGrade(grades: Array(subjects.values))
+            gradeConsole.showAverageGrade(averageGrade)
+        } catch {
+            gradeConsole.printInputError(error)
+        }
+    }
+    
+    func getAverageGrade(grades: [Grade]) throws -> Double {
+        guard !grades.isEmpty else { throw InputError.noGrades }
+        let gradeCount = Double(grades.count)
+        let total = grades.reduce(0) {$0 + $1.asScore}
+        return total / gradeCount
+    }
```
평균을 확인하려면 이름으로 학생을 찾고, 학생의 과목들을 가지고 오고, 성적들을 출력하고, 평균을 계산해서 출력해야하는데요.
평균도 `Student`가 가지고 있으면 어떨까요?
그럼 이름으로 학생을 찾고, 과목을 가져와 성적과 평균을 출력하는 식으로요.

wonhee009 reviewed (changes requested):
@wilybear 스텝2 수고하셨습니다!🎉

많은 고민이 보이는 PR였습니다ㅎㅎ
코드 관련해서 코멘트 달아두었으니 확인 부탁드립니다!

#### Error에 대하여

저도 처음에 커스텀 에러를 만들때 같은 고민을 한적 있는데요.
구조를 잘못 잡으면 커스텀 에러를 100% 활용하기 어렵더라구요.
코드를 작성하기 전에 프로젝트 스펙에 따라 플로우 차트를 먼저 그려보면서 어디서 에러가 발생할 수 있는지에 대해 먼저 체크하면서 작성하니 좀 정리되는 기분이 들더라구요.
이번 프로젝트에서는 할 수 있는만큼 개선을 목표로 하고 이후 프로젝트에서는 에러까지 고려한 설계를 해보시는 것도 좋을거 같습니다.☺️

#### GradeConsoleMessage

저도 관련 Text를 한 곳에 모아두고 사용하는걸 선호하는 편인데요.
이렇게 모아두면 갖는 장점과 단점이 무엇일까요? 같이 얘기해 보는 것도 좋을거 같습니다.😀

#### 고민되는 부분

MARK 주석을 달아두는 것도 하나의 방법이 될 수 있을거 같습니다.
(MARK 주석을 달기 위해서는 관련 있는 것들끼리 뭉쳐두어야 합니다.)
또, property/private property/lifecycle method/method/private method/extension 이렇게 뭉쳐두고 흐름에 따라 정리하는 것도 가독성을 올릴 수 있는 방법이라고 생각합니다.
물론, 이 파일에 이게 있는게 맞나? 에 대한 고민을 먼저 한 후에요☺️

wilybear committed:
Feat: Step2 PR리뷰 반영

- 줄바꿈
- 성적 입력값 검증함수 분리 및 공통 사용 (validateGradeInput)
- 반환 파라미터 이름 부여
- GradeConsoleMessage -> ResultMessage 파일명 수정
- InputError의 localizedError 제대로 활용
- getStudentElement -> getStudent 함수 이름 수정
